### DISCUSSION TITLE
CSINode does not contain driver driver.longhorn.io

### DISCUSSION THREAD

**Discussion by liujiabaoaa** [longhorn/longhorn#6437] [Q&A]
## Question

CSINode izbp196tvao0gtpyqxl8glz does not contain driver driver.longhorn.io

## Environment

QoS Class:                   BestEffort
Node-Selectors:              fua=ljb
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason              Age   From                     Message
  ----     ------              ----  ----                     -------
  Normal   Scheduled           20s   default-scheduler        Successfully assigned es/nginx-d9f54ffbc-q5q6j to izbp196tvao0gtpyqxl8glz
  Warning  FailedAttachVolume  3s    attachdetach-controller  AttachVolume.Attach failed for volume "pvc-66a264c8-53b0-4af1-9f44-cbc922e7190e" : CSINode izbp196tvao0gtpyqxl8glz does not contain driver driver.longhorn.io

## Additional context
想请教下  Node节点有过重启。然后节点上面的pod挂载pvc 失败。报错信息如上

CSINode izbp196tvao0gtpyqxl8glz does not contain driver driver.longhorn.io

我应该如何做？

**Comment by c3y1huang** [comment]
Could you check if the `longhorn-csi-plugin` pods are running and if there are any errors in the logs?

Here is a [similar discussion](https://github.com/longhorn/longhorn/issues/3172).

**Reply by liujiabaoaa** [reply]
1、node-driver-registrar容器的log

[root@iZbp1bxcm25qrv328t3yfmZ ~]# kubectl logs -f longhorn-csi-plugin-22gs2 -c node-driver-registrar -n longhorn-system
I0801 02:37:07.635225    5143 main.go:164] Version: v2.3.0
I0801 02:37:07.635262    5143 main.go:165] Running node-driver-registrar in mode=registration
I0801 02:37:07.636103    5143 main.go:189] Attempting to open a gRPC connection with: "/csi/csi.sock"
I0801 02:37:08.638095    5143 main.go:196] Calling CSI driver to discover driver name
I0801 02:37:08.640312    5143 main.go:206] CSI driver name: "driver.longhorn.io"
I0801 02:37:08.640339    5143 node_register.go:52] Starting Registration Server at: /registration/driver.longhorn.io-reg.sock
I0801 02:37:08.640424    5143 node_register.go:61] Registration Server started at: /registration/driver.longhorn.io-reg.sock
I0801 02:37:08.640487    5143 node_register.go:91] Skipping healthz server because HTTP endpoint is set to: ""



2、longhorn-csi-plugin容器的log

time="2023-08-01T02:37:07Z" level=info msg="Enabling controller service capability: CREATE_DELETE_SNAPSHOT"
time="2023-08-01T02:37:07Z" level=info msg="Enabling controller service capability: CLONE_VOLUME"
time="2023-08-01T02:37:07Z" level=info msg="Enabling volume access mode: SINGLE_NODE_WRITER"
time="2023-08-01T02:37:07Z" level=info msg="Enabling volume access mode: MULTI_NODE_MULTI_WRITER"
time="2023-08-01T02:37:07Z" level=info msg="Listening for connections on address: &net.UnixAddr{Name:\"//csi/csi.sock\", Net:\"unix\"}"
time="2023-08-01T02:37:08Z" level=info msg="GetPluginInfo: req: {}"
time="2023-08-01T02:37:08Z" level=info msg="GetPluginInfo: rsp: {\"name\":\"driver.longhorn.io\",\"vendor_version\":\"v1.2.3\"}"

3、kubelet log日志
Aug 01 19:14:23 izbp196tvao0gtpyqxl8glz kubelet[2170]: E0801 19:14:23.556837    2170 kubelet.go:1720] "Unable to attach or mount volumes for pod; skipping pod" err="unmounted volumes=[elasticsearch-data], unattached volumes=[elastic-internal-transport-certificates downward-api elastic-internal-elasticsearch-config elastic-internal-http-certificates elastic-internal-scripts elastic-internal-elasticsearch-config-local kube-api-access-56wfp elastic-internal-unicast-hosts elasticsearch-data elastic-internal-elasticsearch-plugins-local elasticsearch-logs elastic-internal-probe-user elastic-internal-elasticsearch-bin-local elastic-internal-xpack-file-realm elastic-internal-remote-certificate-authorities]: timed out waiting for the condition" pod="mesoor/data-engine-es-data-nodes-0"

**Reply by c3y1huang** [reply]
>1、node-driver-registrar容器的log
>
>[root@iZbp1bxcm25qrv328t3yfmZ ~]# kubectl logs -f longhorn-csi-plugin-22gs2 -c node-driver-registrar -n longhorn-system
I0801 02:37:07.635225 5143 main.go:164] Version: v2.3.0
I0801 02:37:07.635262 5143 main.go:165] Running node-driver-registrar in mode=registration
I0801 02:37:07.636103 5143 main.go:189] Attempting to open a gRPC connection with: "/csi/csi.sock"
I0801 02:37:08.638095 5143 main.go:196] Calling CSI driver to discover driver name
I0801 02:37:08.640312 5143 main.go:206] CSI driver name: "driver.longhorn.io"
I0801 02:37:08.640339 5143 node_register.go:52] Starting Registration Server at: /registration/driver.longhorn.io-reg.sock
I0801 02:37:08.640424 5143 node_register.go:61] Registration Server started at: /registration/driver.longhorn.io-reg.sock
I0801 02:37:08.640487 5143 node_register.go:91] Skipping healthz server because HTTP endpoint is set to: ""

Seems the longhorn-csi-plugin is missing some expected logs. Could you give [this](https://github.com/longhorn/longhorn/issues/3172#issuecomment-1146385484) a try?

**Reply by liujiabaoaa** [reply]
Can you tell me the file path and file name of this log in the container?

**Reply by liujiabaoaa** [reply]
![Uploading image.png…]()

**Reply by liujiabaoaa** [reply]
My kubectl logs - f only displays so many logs. I don't know the path of the log file in the container

**Reply by c3y1huang** [reply]
When the longhorn-csi-plugin is successfully registered to kubelet, the following logs should be visible:
```
I0728 06:17:07.238374   13523 main.go:102] Received GetInfo call: &InfoRequest{}
I0728 06:17:07.238634   13523 main.go:109] "Kubelet registration probe created" path="/var/lib/kubelet/plugins/driver.longhorn.io/registration"
I0728 06:17:08.668701   13523 main.go:121] Received NotifyRegistrationStatus call: &RegistrationStatus{PluginRegistered:true,Error:,}
```

If these log entries are missing (as based on the node-driver-registrar logs you've provided), it could indicate that the plugin has not been properly registered with kubelet. To address this issue, as suggested in the [discussion](https://github.com/longhorn/longhorn/issues/3172#issuecomment-1146385484), you can try deleting the longhorn-csi-plugin pod to trigger the pod restart. This may help in re-establishing the plugin's registration with kubelet.

**Comment by liujiabaoaa** [comment]
I have already deleted the pods for longhorn csi plugin, and the logs are still the same. Could it be a problem with Kubelet on my node?

**Reply by c3y1huang** [reply]
It's hard to tell without more background context (e.g Longhorn version, Kubernetes version, etc) and logs.

Could you share the output of `kubectl -n longhorn-system get pod`?
For a deeper dive and more thorough analysis, a support-bundle would be super helpful.

**Reply by liujiabaoaa** [reply]
First of all, thank you very much for your patient help.

1. Version

Image: longhornio/longhorn manager: v1.2.3

Client Version: version. Info {Major: "1", Minor: "22", GitVersion: "v1.22.4", GitCommit: "b695d79d4f967c403a96986f1750a35eb75e75f1", GitTreeState: "clean", BuildDate: "2021-11-117T15:48:33Z", GoVersion: "go1.16.10", Compiler: "gc", Platform: "Linux/amd64"}

2、kubectl -n longhorn-system get pod的qin

[root@iZbp1bxcm25qrv328t3yfmZ ~]# kubectl -n longhorn-system get pod
NAME                                        READY   STATUS    RESTARTS       AGE
csi-attacher-5bcbcf4b77-d2vx2               1/1     Running   0              2d6h
csi-attacher-5bcbcf4b77-mz59b               1/1     Running   1 (15d ago)    560d
csi-attacher-5bcbcf4b77-vk6zz               1/1     Running   1 (15d ago)    560d
csi-provisioner-64f66878c7-gmwmb            1/1     Running   0              2d6h
csi-provisioner-64f66878c7-m5gxm            1/1     Running   26 (15d ago)   560d
csi-provisioner-64f66878c7-mm9jp            1/1     Running   1 (15d ago)    560d
csi-resizer-b978fcb49-6mblc                 1/1     Running   1 (15d ago)    560d
csi-resizer-b978fcb49-r8qs4                 1/1     Running   1 (15d ago)    560d
csi-resizer-b978fcb49-tjtdq                 1/1     Running   0              2d6h
csi-snapshotter-78b8849d75-f6d7r            1/1     Running   0              2d6h
csi-snapshotter-78b8849d75-s85wf            1/1     Running   1 (15d ago)    560d
csi-snapshotter-78b8849d75-wlcpw            1/1     Running   1 (15d ago)    560d
engine-image-ei-fa2dfbf0-5pscq              1/1     Running   1 (15d ago)    560d
engine-image-ei-fa2dfbf0-kts84              1/1     Running   0              45h
engine-image-ei-fa2dfbf0-sj964              1/1     Running   1 (15d ago)    560d
instance-manager-e-90f6162d                 1/1     Running   0              22h
instance-manager-e-b2b5e95d                 1/1     Running   0              15d
instance-manager-e-d94f3e91                 1/1     Running   0              15d
instance-manager-r-87dc6262                 1/1     Running   0              15d
instance-manager-r-8970c013                 1/1     Running   0              22h
instance-manager-r-e389ede1                 1/1     Running   0              15d
longhorn-csi-plugin-f9wll                   2/2     Running   0              7h45m
longhorn-csi-plugin-tjj9m                   2/2     Running   2 (15d ago)    560d
longhorn-csi-plugin-w6scq                   2/2     Running   3 (15d ago)    560d
longhorn-driver-deployer-6db849975f-6lrct   1/1     Running   0              2d6h
longhorn-manager-6gz6c                      1/1     Running   0              45h
longhorn-manager-dk5df                      1/1     Running   1 (15d ago)    560d
longhorn-manager-dzvch                      1/1     Running   1 (15d ago)    560d
longhorn-ui-6f547c964-chlfc                 1/1     Running   1 (15d ago)    560d

3、Pod situation of node nodes with no issues


[root@iZbp1bxcm25qrv328t3yfmZ ~]# kubectl -n longhorn-system -o wide get pod|grep  node1
csi-attacher-5bcbcf4b77-d2vx2               1/1     Running   0              2d6h    192.168.38.143   node1                     <none>           <none>
csi-attacher-5bcbcf4b77-vk6zz               1/1     Running   1 (15d ago)    560d    192.168.38.165   node1                     <none>           <none>
csi-provisioner-64f66878c7-gmwmb            1/1     Running   0              2d6h    192.168.38.131   node1                     <none>           <none>
csi-provisioner-64f66878c7-mm9jp            1/1     Running   1 (15d ago)    560d    192.168.38.182   node1                     <none>           <none>
csi-resizer-b978fcb49-r8qs4                 1/1     Running   1 (15d ago)    560d    192.168.38.158   node1                     <none>           <none>
csi-resizer-b978fcb49-tjtdq                 1/1     Running   0              2d6h    192.168.38.151   node1                     <none>           <none>
csi-snapshotter-78b8849d75-f6d7r            1/1     Running   0              2d6h    192.168.38.134   node1                     <none>           <none>
csi-snapshotter-78b8849d75-wlcpw            1/1     Running   1 (15d ago)    560d    192.168.38.177   node1                     <none>           <none>
engine-image-ei-fa2dfbf0-sj964              1/1     Running   1 (15d ago)    560d    192.168.38.174   node1                     <none>           <none>
instance-manager-e-b2b5e95d                 1/1     Running   0              15d     192.168.38.153   node1                     <none>           <none>
instance-manager-r-87dc6262                 1/1     Running   0              15d     192.168.38.148   node1                     <none>           <none>
longhorn-csi-plugin-tjj9m                   2/2     Running   2 (15d ago)    560d    192.168.38.188   node1                     <none>           <none>
longhorn-driver-deployer-6db849975f-6lrct   1/1     Running   0              2d6h    192.168.38.133   node1                     <none>           <none>
longhorn-manager-dzvch                      1/1     Running   1 (15d ago)    560d    192.168.38.183   node1                     <none>           <none>
longhorn-ui-6f547c964-chlfc                 1/1     Running   1 (15d ago)    560d    192.168.38.157   node1                     <none>           <none>

4、Pod situation of problematic node nodes

[root@iZbp1bxcm25qrv328t3yfmZ ~]# kubectl -n longhorn-system -o wide get pod|grep izbp196tvao0gtpyqxl8glz
engine-image-ei-fa2dfbf0-kts84              1/1     Running   0              45h     192.168.17.130   izbp196tvao0gtpyqxl8glz   <none>           <none>
instance-manager-e-90f6162d                 1/1     Running   0              22h     192.168.17.138   izbp196tvao0gtpyqxl8glz   <none>           <none>
instance-manager-r-8970c013                 1/1     Running   0              22h     192.168.17.139   izbp196tvao0gtpyqxl8glz   <none>           <none>
longhorn-csi-plugin-f9wll                   2/2     Running   0              7h48m   192.168.17.140   izbp196tvao0gtpyqxl8glz   <none>           <none>
longhorn-manager-6gz6c                      1/1     Running   0              45h     192.168.17.131   izbp196tvao0gtpyqxl8glz   <none>           <none>

**Reply by c3y1huang** [reply]
It doesn't look like there is anything out of the ordinary apart from the cluster using a quite old version of Longhorn and Kubernetes. Do you have the support-bundle?

Maybe you can try to deploy a new node and then migrate the workloads over.

Just curious, any particular reason you are sticking to those versions?
